File: winning-bd-complete/winningbdt-final/android-app/app/src/main/java/com/winningbd/admin/service/SmsBridgeService.java
package com.winningbd.admin.service;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.IBinder;
import androidx.core.app.NotificationCompat;
import com.winningbd.admin.R;

public class SmsBridgeService extends Service {

    private static final String CHANNEL_ID = "winningbd_sms_channel";
    private static final int NOTIFICATION_ID = 1001;
    private static boolean started = false;

    public static void ensureStarted(Context context) {
        if (started) return;
        Intent intent = new Intent(context, SmsBridgeService.class);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            context.startForegroundService(intent);
        } else {
            context.startService(intent);
        }
        started = true;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        createNotificationChannel();
        Notification notification = buildNotification();
        startForeground(NOTIFICATION_ID, notification);
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        return START_STICKY;
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    private void createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel channel = new NotificationChannel(
                CHANNEL_ID,
                "SMS Payment Checker",
                NotificationManager.IMPORTANCE_LOW
            );
            channel.setDescription("Background service for SMS payment verification");
            NotificationManager manager = getSystemService(NotificationManager.class);
            if (manager != null) {
                manager.createNotificationChannel(channel);
            }
        }
    }

    private Notification buildNotification() {
        Intent intent = new Intent(this, com.winningbd.admin.web.MainActivity.class);
        PendingIntent pendingIntent = PendingIntent.getActivity(
            this, 0, intent, PendingIntent.FLAG_IMMUTABLE
        );
        
        return new NotificationCompat.Builder(this, CHANNEL_ID)
            .setContentTitle("WINNING BD Admin")
            .setContentText("SMS Payment Checker running")
            .setSmallIcon(R.drawable.ic_notification)
            .setContentIntent(pendingIntent)
            .setPriority(NotificationCompat.PRIORITY_LOW)
            .setCategory(NotificationCompat.CATEGORY_SERVICE)
            .setOngoing(true)
            .build();
    }
}

File: android-app/app/src/main/java/com/winningbd/admin/service/SmsBridgeService.kt
package com.winningbd.admin.service

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.Service
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Handler
import android.os.IBinder
import android.os.Looper
import androidx.core.app.NotificationCompat
import com.winningbd.admin.R
import com.winningbd.admin.store.Prefs

/**
 * Foreground "dataSync" service. Keeps the SMS bridge alive and flushes the
 * offline queue periodically. Started on boot, app-open and connectivity
 * changes; the workers catch the Android 12+ background-start restrictions
 * and fall back to inline flushing.
 */
class SmsBridgeService : Service() {

    private val handler = Handler(Looper.getMainLooper())
    private val flushEveryMs = 300_000L

    private val flushRunnable = object : Runnable {
        override fun run() {
            QueueFlusher.flushAsync(this@SmsBridgeService)
            handler.postDelayed(this, flushEveryMs)
        }
    }

    override fun onBind(intent: Intent?): IBinder? = null

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        try {
            startForeground(1, buildNotification())
        } catch (e: Exception) {
            stopSelf()
            return START_NOT_STICKY
        }
        QueueFlusher.flushAsync(this)
        handler.removeCallbacks(flushRunnable)
        handler.postDelayed(flushRunnable, flushEveryMs)
        return START_STICKY
    }

    override fun onDestroy() {
        handler.removeCallbacks(flushRunnable)
        super.onDestroy()
    }

    private fun buildNotification(): Notification {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val channel = NotificationChannel(
                CHANNEL_ID,
                getString(R.string.channel_sms),
                NotificationManager.IMPORTANCE_LOW
            )
            val nm = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            nm.createNotificationChannel(channel)
        }
        val prefs = Prefs(this)
        val text = if (prefs.deviceId.isEmpty()) {
            getString(R.string.service_notif_text)
        } else {
            "Device ${prefs.deviceId.take(14)}\u2026 syncing"
        }
        return NotificationCompat.Builder(this, CHANNEL_ID)
            .setContentTitle(getString(R.string.service_notif_title))
            .setContentText(text)
            .setSmallIcon(R.drawable.ic_notification)
            .setOngoing(true)
            .setPriority(NotificationCompat.PRIORITY_LOW)
            .build()
    }

    companion object {
        private const val CHANNEL_ID = "sms_bridge"

        fun ensureStarted(context: Context) {
            val app = context.applicationContext
            try {
                app.startService(Intent(app, SmsBridgeService::class.java))
            } catch (e: Exception) {
                // Background FGS start may be restricted — inline flush instead.
                QueueFlusher.flushAsync(app)
            }
        }
    }
}